"Found multiple definitions for subrepo <arch>" error when cross-compiling

Spotted today in the context of the [go-rules release workflow](https://github.com/please-build/go-rules/actions/runs/10388260693/job/28783305957), but has been seen before too:

```
$ ./pleasew build --profile release -o plugin.go.pleasegotool:$(pwd)/plz-out/bin/tools/please_go/please_go -p -v notice //package:please_go_release_files //package:driver_release_files
20:00:09.384 NOTICE: Build running for 0s, 0 / 1 tasks done, 0 workers busy
20:00:09.407 CRITICAL: Found multiple definitions for subrepo 'darwin_arm64' (&{Name:darwin_arm64 Root: PackageRoot: Target:<nil> State:0xc00021a3c0 Arch:{OS:darwin Arch:arm64} IsCrossCompile:true AdditionalConfigFiles:[] fs:{} fsSync:{done:{_:{} v:1} m:{state:0 sema:0}}} s &{Name:darwin_arm64 Root: PackageRoot: Target:<nil> State:0xc00021a3c0 Arch:{OS:darwin Arch:arm64} IsCrossCompile:true AdditionalConfigFiles:[] fs:<nil> fsSync:{done:{_:{} v:0} m:{state:0 sema:0}}})
Error: Process completed with exit code 1.
```

The architecture subrepo causing the problem isn't consistent - I've seen it fail with `darwin_arm64` and `linux_amd64` on different runs of the workflow. From the function that produces the error message, it seems that two subrepos can be equal if the structs representing them are deeply equal:

https://github.com/thought-machine/please/blob/382cfd6f8ab1bb08d70ef150f963d1a00c848847/src/core/graph.go#L105-L114

Perhaps this condition is slightly too strict - the unexported `fs` and `fsSync` fields were added relatively recently, in #3038, and from the error message above it appears that the subrepos are identical in other respects, so perhaps they should be excluded from the equality comparison.

--- a/go.mod
+++ b/go.mod
@@ -14,6 +14,7 @@ require (
 	github.com/dustin/go-humanize v1.0.1
 	github.com/fsnotify/fsnotify v1.7.0
 	github.com/golang/protobuf v1.5.4
+	github.com/google/go-cmp v0.6.0
 	github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
 	github.com/google/uuid v1.6.0
 	github.com/grpc-ecosystem/go-grpc-middleware v1.4.0

--- a/src/core/BUILD
+++ b/src/core/BUILD
@@ -21,6 +21,8 @@ go_library(
         "///third_party/go/github.com_thought-machine_go-flags//:go-flags",
         "///third_party/go/github.com_zeebo_blake3//:blake3",
         "///third_party/go/golang.org_x_sync//errgroup",
+        "///third_party/go/github.com_google_go-cmp//cmp",
+        "///third_party/go/github.com_google_go-cmp//cmp/cmpopts",
         "//src/cli",
         "//src/cli/logging",
         "//src/cmap",

--- a/src/core/graph.go
+++ b/src/core/graph.go
@@ -5,7 +5,6 @@
 package core
 
 import (
-	"reflect"
 	"sort"
 
 	"github.com/thought-machine/please/src/cmap"
@@ -105,7 +104,7 @@ func (graph *BuildGraph) AddSubrepo(subrepo *Subrepo) {
 func (graph *BuildGraph) MaybeAddSubrepo(subrepo *Subrepo) *Subrepo {
 	if !graph.subrepos.Add(subrepo.Name, subrepo) {
 		old := graph.subrepos.Get(subrepo.Name)
-		if !reflect.DeepEqual(old, subrepo) {
+		if !old.Equal(subrepo) {
 			log.Fatalf("Found multiple definitions for subrepo '%s' (%+v s %+v)", old.Name, old, subrepo)
 		}
 		return old

--- a/src/core/subrepo.go
+++ b/src/core/subrepo.go
@@ -8,10 +8,10 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/thought-machine/please/src/cli"
-	"github.com/thought-machine/please/src/fs"
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/thought-machine/please/src/cli"
+	"github.com/thought-machine/please/src/fs"
 )
 
 // A Subrepo stores information about a registered subrepository, typically one

--- a/src/core/subrepo.go
+++ b/src/core/subrepo.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
+
 	"github.com/thought-machine/please/src/cli"
 	"github.com/thought-machine/please/src/fs"
 )

--- a/src/core/subrepo.go
+++ b/src/core/subrepo.go
@@ -10,6 +10,8 @@ import (
 
 	"github.com/thought-machine/please/src/cli"
 	"github.com/thought-machine/please/src/fs"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
 )
 
 // A Subrepo stores information about a registered subrepository, typically one
@@ -66,6 +68,11 @@ func (s *Subrepo) FS() iofs.FS {
 // IsRemoteSubrepo returns true when the subrepo sources are remote i.e. not downloaded to plz-out
 func (s *Subrepo) IsRemoteSubrepo() bool {
 	return s.Root != "" && s.Target != nil && !s.Target.Local && s.State.RemoteClient != nil
+}
+
+// Equal returns true if this subrepo is equivalent to another, or false if it is not.
+func (s *Subrepo) Equal(other *Subrepo) bool {
+	return cmp.Equal(s, other, cmpopts.IgnoreFields(Subrepo{}, "fs", "fsSync"))
 }
 
 // SubrepoForArch creates a new subrepo for the given architecture.

